src/spyglass/common/common_interval.py

```diff
@@ -209,6 +225,537 @@ def cleanup(self, dry_run=True):
+                if intervals1[i][0] < intervals2[j][0]:
+                    result.append((intervals1[i][0], intervals2[j][0]))
+                if intervals1[i][1] > intervals2[j][1]:
+                    intervals1[i] = (intervals2[j][1], intervals1[i][1])
+                    j += 1
+                else:
+                    i += 1
+        result += intervals1[i:]
+        return Interval(result)
+
+    def complement(self, other, min_length=0.0):
+        """
+        Finds intervals in intervals1 that are not in intervals2
+
+        Parameters
+        ----------
+        min_length : float, optional
+            Minimum interval length in seconds. Defaults to 0.0.
+        """
+        intervals1 = self.times
+        intervals2 = self._extract(other)
+
+        result = []
+
+        for start1, end1 in intervals1:
+            subtracted = [(start1, end1)]
+
+            for start2, end2 in intervals2:
+                new_subtracted = []
+
+                for s, e in subtracted:
+                    if start2 <= s and e <= end2:
+                        continue
+                    if e <= start2 or end2 <= s:
+                        new_subtracted.append((s, e))
+                        continue
+                    if start2 > s:
+                        new_subtracted.append((s, start2))
+                    if end2 < e:
+                        new_subtracted.append((end2, e))
+
+                subtracted = new_subtracted
+
+            result.extend(subtracted)
+
+        return Interval(
+            self._by_length(
+                np.asarray(result), min_length=min_length, max_length=1e100
+            )
+        )
+
+    def add_removal_window(self, removal_window_ms, timestamps):
+        """Add half of a removal window to start/end of each interval"""
+        half_win = removal_window_ms / 1000 * 0.5
+        new_interval = np.zeros((len(self.times), 2), dtype=np.float64)
+        # for each interval, add half removal window to start and end
+        for idx, interval in enumerate(self.times):
+            start, end = interval[0], interval[1]
+            new_interval[idx] = [
+                timestamps[start] - half_win,
+                np.minimum(timestamps[end] + half_win, timestamps[-1]),
+            ]
+        # NOTE: np.min and if len cond were in
+        # spikesorting_artifact, but not lfp_artifact
+        return (
+            self._union_consolidate(new_interval)
+            if len(new_interval) > 1
+            else Interval(new_interval)
+        )
+
+    def to_indices(self, timestamps):
+        """Convert intervals to indices in the given timestamps.
+
+        Parameters
+        ----------
+        timestamps : array_like
+            The timestamps to convert the intervals to indices in.
+
+        Returns
+        -------
+        np.ndarray
+            The indices of the intervals in the given timestamps.
+        """
+        starts = np.searchsorted(timestamps, self.times[:, 0])
```

Comment:
would probably be even more efficient with one searchsorted and a reshape

Reply:
I'm not sure what you had in mind

Reply:
It's not a big deal but:
`np.searchsorted(timestamps, self.times.ravel()).reshape(-1, 2)`

I think would work?

Reply:
It matches my existing with execution time within a few ms. It looks like there's some variation in the types returned by the yours+mine vs the original: `array([[int,int],[int,int]])` vs `[array([int,int]),array([int,int])]`, but the only situations that use this then apply `add_removal_window`, which iterates over them